File: app/api/ai_scoring.py
"""AI 评分路由：标签偏好 CRUD/补全 + 组合 AI 打分 + 每日 AI 打分。

- 偏好：用户输入简短偏好，保存时自动 AI 补全（draft），确认后（confirmed）用于打分。
- 组合：POST 手动触发 AI 打分（可带 tags 标签筛选）；GET 读最新报告（含 stale）。
- 每日：POST 某日 AI 打分（一次调用逐笔+汇总）；GET 读该日详情与报告。
- GET 全部零网络零写；AI POST 无激活模型返回 400。
"""
import logging

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app.services import ai as ai_svc
from app.services import ai_scoring as svc

logger = logging.getLogger("api")
router = APIRouter()


class TagPrefBody(BaseModel):
    raw_pref: str
    prompt: str | None = None
    auto_expand: bool = True


class ExpandBody(BaseModel):
    raw_pref: str


class DateBody(BaseModel):
    date: str


def _parse_tags(tags: str | None) -> list[str] | None:
    """逗号分隔的标签筛选 → list；空/None → None（全部）。"""
    if not tags:
        return None
    out = [t.strip() for t in tags.split(",") if t.strip()]
    return out or None


def _configured() -> bool:
    return ai_svc.get_active_model() is not None


# ============================================================ 标签偏好

@router.get("/ai-scoring/prefs")
def list_prefs():
    """全部标签偏好 + 是否已配置 AI。"""
    prefs = svc.list_tag_prefs()
    for p in prefs:
        p["status_cn"] = "已确认" if p["status"] == "confirmed" else "待确认"
    return {"ok": True, "data": {"prefs": prefs, "configured": _configured()}}


@router.get("/ai-scoring/prefs/{tag}")
def get_pref(tag: str):
    return {"ok": True, "data": svc.get_tag_pref(tag)}


@router.put("/ai-scoring/prefs/{tag}")
def save_pref(tag: str, body: TagPrefBody):
    """保存偏好。有 prompt → 直接存（confirmed）；否则 auto_expand 且已配置 → AI 补全存 draft；
    AI 失败不阻断保存（返回行 + expand_error）。"""
    try:
        if body.prompt is not None:
            row = svc.upsert_tag_pref(tag, body.raw_pref, prompt=body.prompt)
            expanded, error = False, None
        elif body.auto_expand and _configured():
            try:
                row = svc.expand_tag_prompt(tag, body.raw_pref)
                expanded, error = True, None
            except ValueError as e:
                row = svc.upsert_tag_pref(tag, body.raw_pref)
                expanded, error = False, str(e)
        else:
            row = svc.upsert_tag_pref(tag, body.raw_pref)
            expanded, error = False, None
    except ValueError as e:
        raise HTTPException(400, str(e))
    data = dict(row)
    data["expanded"] = expanded
    data["expand_error"] = error
    logger.info("[标签偏好] 保存 %s → %s", tag, row["status"])
    return {"ok": True, "data": data}


@router.post("/ai-scoring/prefs/{tag}/expand")
def expand_pref(tag: str, body: ExpandBody):
    """手动「AI 补全」：生成完整评分指引存 draft（待确认）。"""
    try:
        row = svc.expand_tag_prompt(tag, body.raw_pref)
    except ValueError as e:
        raise HTTPException(400, str(e))
    return {"ok": True, "data": row}


@router.post("/ai-scoring/prefs/{tag}/confirm")
def confirm_pref(tag: str):
    """确认偏好生效（仅 confirmed 用于打分）。"""
    try:
        row = svc.confirm_tag_pref(tag)
    except ValueError as e:
        raise HTTPException(400, str(e))
    return {"ok": True, "data": row}


@router.delete("/ai-scoring/prefs/{tag}")
def remove_pref(tag: str):
    svc.delete_tag_pref(tag)
    return {"ok": True, "data": {"deleted": tag}}


# ============================================================ 组合 AI

@router.get("/ai-scoring/portfolio")
def portfolio_report(tags: str | None = None):
    """读取最新组合 AI 报告（含 stale：组合已变化建议重新打分）。纯读零网络零写。"""
    return {"ok": True, "data": {
        "report": svc.get_portfolio_report(_parse_tags(tags)),
        "configured": _configured(),
    }}


@router.post("/ai-scoring/portfolio")
def score_portfolio(tags: str | None = None):
    """手动触发组合 AI 打分（跟随标签筛选：?tags=红利,科技）。"""
    try:
        result = svc.score_portfolio(_parse_tags(tags))
    except ValueError as e:
        raise HTTPException(400, str(e))
    logger.info("[AI打分] 组合%s 完成 %s（%s）", f"筛选{tags}" if tags else "全部", result["report"]["score"], result["report"]["rating"])
    return {"ok": True, "data": result}


# ============================================================ 每日 AI

@router.get("/ai-scoring/daily-reports")
def daily_reports():
    """左目录：所有 buy/sell 交易日（倒序），各附 AI 报告摘要（无则 ai=null）。纯读。"""
    return {"ok": True, "data": {"days": svc.list_daily_days(), "configured": _configured()}}


@router.get("/ai-scoring/daily")
def daily_report(date: str):
    """某日详情：交易表 + AI 报告（无报告则 report=null）。纯读零网络零写。"""
    return {"ok": True, "data": {
        "configured": _configured(),
        "day": svc.get_daily_day(date),
        "report": (svc.get_daily_report(date) or {}).get("report"),
    }}


@router.post("/ai-scoring/daily")
def score_daily(body: DateBody):
    """某日 AI 打分（一次调用逐笔+汇总）。无交易或无激活模型返回 400。"""
    try:
        result = svc.score_daily(body.date)
    except ValueError as e:
        raise HTTPException(400, str(e))
    if result is None:
        raise HTTPException(400, "该日无交易")
    logger.info("[AI打分] 当日 %s 完成 %s（%s）", body.date, result["report"]["score"], result["report"]["rating"])
    return {"ok": True, "data": {"day": svc.get_daily_day(body.date), "report": result["report"]}}

File: backend/internal/route/ai_scoring.go
package route

// AI 评分路由：标签偏好 / 组合打分 / 每日打分。对齐 app/api/ai_scoring.py。

import (
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"

	"stockanalyzer/internal/service/ai"
	"stockanalyzer/internal/service/jobs"
)

func setupAIScoringRoutes(api *gin.RouterGroup, s *Services, aiSvc *ai.Service) {
	parseTags := func(c *gin.Context) []string {
		qs := strings.TrimSpace(c.Query("tags"))
		if qs == "" {
			return nil // 对齐 Python：无参数 = None = 全部（nil 与空切片语义不同）
		}
		var out []string
		for _, t := range strings.Split(qs, ",") {
			if t = strings.TrimSpace(t); t != "" {
				out = append(out, t)
			}
		}
		return out
	}
	configured := func() bool { return aiSvc.GetActiveModel() != nil }

	// ---- 标签偏好 ----
	// GET /api/ai-scoring/prefs —— 列出全部标签偏好（对齐 app/api/ai_scoring.py）：返回 prefs 列表（附带 status_cn 已确认/待确认）
	// 与 configured（是否已配置模型）。
	api.GET("/ai-scoring/prefs", func(c *gin.Context) {
		prefs := aiSvc.ListTagPrefs()
		for _, p := range prefs {
			if st, _ := p["status"].(string); st == "confirmed" {
				p["status_cn"] = "已确认"
			} else {
				p["status_cn"] = "待确认"
			}
		}
		c.JSON(http.StatusOK, gin.H{"ok": true, "data": gin.H{"prefs": prefs, "configured": configured()}})
	})
	// GET /api/ai-scoring/prefs/:tag —— 读取单个标签偏好（对齐 app/api/ai_scoring.py）：返回 draft/confirmed 指引。
	api.GET("/ai-scoring/prefs/:tag", func(c *gin.Context) {
		c.JSON(http.StatusOK, gin.H{"ok": true, "data": aiSvc.GetTagPref(c.Param("tag"))})
	})
	// PUT /api/ai-scoring/prefs/:tag —— 保存标签偏好（对齐 app/api/ai_scoring.py）：body 传 raw_pref；prompt 非空则直接存完整指引，
	// auto_expand=true 且已配模型时先请求 AI 补全（draft），补全失败回退原样保存；返回 row + expanded/expand_error。
	api.PUT("/ai-scoring/prefs/:tag", func(c *gin.Context) {
		tag := c.Param("tag")
		var body struct {
			RawPref    string  `json:"raw_pref"`
			Prompt     *string `json:"prompt"`
			AutoExpand bool    `json:"auto_expand"`
		}
		_ = c.ShouldBindJSON(&body)
		var row map[string]any
		expanded := false
		var expandErr any
		var err error
		if body.Prompt != nil {
			row, err = aiSvc.UpsertTagPref(tag, body.RawPref, *body.Prompt, "")
		} else if body.AutoExpand && configured() {
			row, err = aiSvc.ExpandTagPrompt(tag, body.RawPref)
			expanded = true
			if err != nil {
				row, err = aiSvc.UpsertTagPref(tag, body.RawPref, "", "")
				expanded = false
				expandErr = err.Error()
			}
		} else {
			row, err = aiSvc.UpsertTagPref(tag, body.RawPref, "", "")
		}
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"detail": err.Error()})
			return
		}
		if row == nil {
			c.JSON(http.StatusBadRequest, gin.H{"detail": "保存失败"})
			return
		}
		row["expanded"] = expanded
		row["expand_error"] = expandErr
		c.JSON(http.StatusOK, gin.H{"ok": true, "data": row})
	})
	// POST /api/ai-scoring/prefs/:tag/expand —— 手动发起该标签偏好的 AI 补全（对齐 app/api/ai_scoring.py）：body 传 raw_pref，返回补全后的 draft 指引。
	api.POST("/ai-scoring/prefs/:tag/expand", func(c *gin.Context) {
		var body struct {
			RawPref string `json:"raw_pref"`
		}
		_ = c.ShouldBindJSON(&body)
		row, err := aiSvc.ExpandTagPrompt(c.Param("tag"), body.RawPref)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"detail": err.Error()})
			return
		}
		c.JSON(http.StatusOK, gin.H{"ok": true, "data": row})
	})
	// POST /api/ai-scoring/prefs/:tag/confirm —— 确认标签偏好（draft → confirmed，对齐 app/api/ai_scoring.py），确认后才用于打分。
	api.POST("/ai-scoring/prefs/:tag/confirm", func(c *gin.Context) {
		row, err := aiSvc.ConfirmTagPref(c.Param("tag"))
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"detail": err.Error()})
			return
		}
		c.JSON(http.StatusOK, gin.H{"ok": true, "data": row})
	})
	// DELETE /api/ai-scoring/prefs/:tag —— 删除标签偏好（对齐 app/api/ai_scoring.py）：返回被删除的 tag。
	api.DELETE("/ai-scoring/prefs/:tag", func(c *gin.Context) {
		tag := c.Param("tag")
		_ = aiSvc.DeleteTagPref(tag)
		c.JSON(http.StatusOK, gin.H{"ok": true, "data": gin.H{"deleted": tag}})
	})

	// ---- 组合打分 ----
	// GET /api/ai-scoring/portfolio —— 读取组合打分报告（对齐 app/api/ai_scoring.py）：query tags 逗号分隔筛选子组合，
	// 空=「全部」；返回 report + configured；画像变化标 stale。
	api.GET("/ai-scoring/portfolio", func(c *gin.Context) {
		tags := parseTags(c)
		c.JSON(http.StatusOK, gin.H{"ok": true, "data": gin.H{
			"report": aiSvc.GetPortfolioReport(tags), "configured": configured(),
		}})
	})
	// POST /api/ai-scoring/portfolio —— 触发组合 AI 打分（异步，对齐 app/api/ai_scoring.py）：query tags 筛选子组合；
	// body 可传 system_prompt/intensity；未配置模型 400；返回 job_id。
	api.POST("/ai-scoring/portfolio", func(c *gin.Context) {
		if !configured() {
			c.JSON(http.StatusBadRequest, gin.H{"detail": "未配置 AI 模型"})
			return
		}
		tags := parseTags(c)
		var body struct {
			SystemPrompt *string `json:"system_prompt"`
			Intensity    string  `json:"intensity"`
		}
		_ = c.ShouldBindJSON(&body)
		intensity := body.Intensity
		if intensity == "" {
			intensity = "normal"
		}
		var prompt string
		if body.SystemPrompt != nil {
			prompt = *body.SystemPrompt
		}
		label := "组合 AI 打分"
		if len(tags) > 0 {
			label += "·" + strings.Join(tags, ",")
		}
		jobID := s.Jobs.Start("ai.portfolio", label, func(p *jobs.Progress) error {
			_, err := aiSvc.ScorePortfolio(tags, prompt, intensity)
			return err
		})
		c.JSON(http.StatusOK, gin.H{"ok": true, "data": gin.H{"job_id": jobID, "async": true}})
	})

	// ---- 每日打分 ----
	// GET /api/ai-scoring/daily-reports —— 列出有每日打分的日期（对齐 app/api/ai_scoring.py）：返回 days + configured。
	api.GET("/ai-scoring/daily-reports", func(c *gin.Context) {
		c.JSON(http.StatusOK, gin.H{"ok": true, "data": gin.H{
			"days": aiSvc.ListDailyDays(), "configured": configured(),
		}})
	})
	// GET /api/ai-scoring/daily —— 读取某日每日打分报告（对齐 app/api/ai_scoring.py）：query date；返回 configured/day/report。
	api.GET("/ai-scoring/daily", func(c *gin.Context) {
		date := c.Query("date")
		var report any
		if r := aiSvc.GetDailyReport(date); r != nil {
			report = r["report"]
		}
		c.JSON(http.StatusOK, gin.H{"ok": true, "data": gin.H{
			"configured": configured(), "day": aiSvc.GetDailyDay(date), "report": report,
		}})
	})
	// POST /api/ai-scoring/daily —— 触发某日每日 AI 打分（异步，对齐 app/api/ai_scoring.py）：
	// body 需 score_date（必填），可传 system_prompt/intensity；未配置模型 400；返回 job_id。
	api.POST("/ai-scoring/daily", func(c *gin.Context) {
		if !configured() {
			c.JSON(http.StatusBadRequest, gin.H{"detail": "未配置 AI 模型"})
			return
		}
		var body struct {
			ScoreDate    string  `json:"score_date"`
			SystemPrompt *string `json:"system_prompt"`
			Intensity    string  `json:"intensity"`
		}
		if err := c.ShouldBindJSON(&body); err != nil || body.ScoreDate == "" {
			c.JSON(http.StatusBadRequest, gin.H{"detail": "缺少 score_date"})
			return
		}
		intensity := body.Intensity
		if intensity == "" {
			intensity = "normal"
		}
		var prompt string
		if body.SystemPrompt != nil {
			prompt = *body.SystemPrompt
		}
		jobID := s.Jobs.Start("ai.daily", "每日 AI 打分 "+body.ScoreDate, func(p *jobs.Progress) error {
			_, err := aiSvc.ScoreDaily(body.ScoreDate, prompt, intensity)
			return err
		})
		c.JSON(http.StatusOK, gin.H{"ok": true, "data": gin.H{"job_id": jobID, "async": true}})
	})
}
